overlord/snapstate/snapstate.go

```diff
@@ -3502,14 +3528,49 @@ func removeInactiveRevision(st *state.State, name, snapID string, revision snap.
 		// no version info needed
 	}
 
-	clearData := st.NewTask("clear-snap", fmt.Sprintf(i18n.G("Remove data for snap %q (%s)"), name, revision))
+	clearData := st.NewTask("clear-snap",
+		fmt.Sprintf(i18n.G("Remove data for snap %q (%s)"), name, revision))
 	clearData.Set("snap-setup", snapsup)
+	tasks = append(tasks, clearData)
 
-	discardSnap := st.NewTask("discard-snap", fmt.Sprintf(i18n.G("Remove snap %q (%s) from the system"), name, revision))
-	discardSnap.WaitFor(clearData)
+	// Discard components first
+	cinfos, err := snapst.ComponentInfosForRevision(revision)
+	if err != nil {
+		return nil, err
+	}
+	discardCompTasks := make([]*state.Task, len(cinfos))
+	for i, cinfo := range cinfos {
+		compsup := ComponentSetup{
+			CompSideInfo: &cinfo.ComponentSideInfo,
+			CompType:     cinfo.Type,
+			componentInstallFlags: componentInstallFlags{
+				SkipProfiles: true,
+			},
+		}
+
+		discardComp := st.NewTask("discard-component", fmt.Sprintf(i18n.G(
+			"Discard revision for component %q"),
+			compsup.CompSideInfo.Component))
+		discardComp.Set("snap-setup-task", clearData.ID())
+		discardComp.Set("component-setup", compsup)
+		compState := sequence.NewComponentState(&cinfo.ComponentSideInfo, cinfo.Type)
+		discardComp.Set("unlinked-component", compState)
+		discardComp.WaitFor(clearData)
+		discardCompTasks[i] = discardComp
+		tasks = append(tasks, discardComp)
```

Comment:
Why does `discard-component` need a `sequence.ComponentState` attached to it? It looks like the task only actually uses the `snap.ComponentSideInfo` from it, which also exists in the `ComponentSetup`, which is also attached to the task.

Reply:
`discard-component` might be used also when refreshing just components, and it that case the `ComponentSetup` contains the new revision to install, while `doUnlinkCurrentComponent` stores the just unlinked component in `unlinked-component`. So we cannot use `ComponentSetup` in all cases.

Reply:
But you had a good point here as depending on whether the revision in both compSetup and the unlinked component are the same (snap being removed) or not (component being replaced) we should try to remove or not the symlink to the active component. So I've changed that in https://github.com/snapcore/snapd/pull/14235/commits/8b1f8ee5348e1b6cd43aa403e78153750994e934.